src/main/java/dev/morling/onebrc/CalculateAverage_michaljonko.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    private static String sortedResults(Collection<StationMeasurement> results) {
+        return results.stream()
+                .sorted(Comparator.comparing(stationMeasurement -> stationMeasurement.station.name()))
+                .map(StationMeasurement::data)
+                .collect(Collectors.joining(", ", "{", "}"));
+    }
+
+    private static Collection<StationMeasurement> calculate(Path path, int partitionsAmount) {
+        try (var parseExecutorService = Executors.newFixedThreadPool(partitionsAmount);
+                var mergeExecutorService = Executors.newSingleThreadExecutor()) {
+            final var memorySegments = FilePartitioner.createSegments(path, partitionsAmount);
+            final var parseFutures = new ArrayList<Future<HashMap<Integer, StationMeasurement>>>(memorySegments.size());
+            final var results = new HashMap<Station, StationMeasurement>();
+            for (var memorySegment : memorySegments) {
+                parseFutures.add(parseExecutorService.submit(() -> parse(memorySegment)));
+            }
+            Future<?> lastFuture = null;
+            for (var future : parseFutures) {
+                final var futureResult = future.get();
+                lastFuture = mergeExecutorService.submit(
+                        () -> futureResult.forEach((k, v) -> results.merge(v.station, v, StationMeasurement::update)));
+            }
+            lastFuture.get();
+            return results.values();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static HashMap<Integer, StationMeasurement> parse(MemorySegment memorySegment) {
+        final var stationNameRaw = new byte[MAX_STATION_NAME_LENGTH];
+        final var temperatureRaw = new byte[MAX_TEMPERATURE_LENGTH];
+        final var memorySegmentSize = memorySegment.byteSize();
+        final var stationsMap = new HashMap<Integer, StationMeasurement>(MAX_STATION_NAMES, 1f);
+
+        var stationNameRawIndex = 0;
+        var stationNameHash = 0;
+        var temperatureRawIndex = 0;
+        var memorySegmentOffset = 0L;
+
+        byte b;
+        StationMeasurement stationMeasurement;
+
+        while (memorySegmentOffset < memorySegmentSize) {
+            while ((b = memorySegment.get(ValueLayout.JAVA_BYTE, memorySegmentOffset++)) != ';') {
+                stationNameRaw[stationNameRawIndex++] = b;
+                stationNameHash = 31 * stationNameHash + b;
+            }
+
+            while (memorySegmentOffset < memorySegmentSize && (b = memorySegment.get(ValueLayout.JAVA_BYTE, memorySegmentOffset++)) != '\n') {
+                temperatureRaw[temperatureRawIndex++] = b;
+            }
+
+            final var temperature = TemperatureParser.parse(temperatureRaw, temperatureRawIndex);
+            stationMeasurement = stationsMap.get(stationNameHash);
+            if (stationMeasurement != null) {
```

Comment:
I think that if 2 different station names would have the same hash, the first in the map would get the temperature of the other station.

Reply:
Hum, yeah, indeed. You can't condense a 100 bytes string to an integer in a map without losing information.

Reply:
For a case with known 10k cities (`weather_stations.csv`) it works. The last commits introduced a solution to minimize collisions when 10000 different city names are used.